src/java.net.http/share/classes/jdk/internal/net/http/Stream.java

```diff
@@ -565,48 +577,46 @@ request, exchange, responseHeaders, connection(),
 
     }
 
+    // Logic here starts checking roughly in reverse order of the Stream's life-cycle i.e. check if closed, check if
+    // closing or done sending, response received or waiting for expect continue, receiving reset on active stream...
     void incoming_reset(ResetFrame frame) {
         Log.logTrace("Received RST_STREAM on stream {0}", streamid);
-        if (endStreamReceived()) {
+        Flow.Subscriber<?> subscriber = responseSubscriber == null ? pendingResponseSubscriber : responseSubscriber;
```

Comment:
```suggestion
        Flow.Subscriber<?> subscriber = responseSubscriber;
        if (subscriber == null) subscriber = pendingResponseSubscriber;
```
(avoids reading volatile twice)

Reply:
Ah yes, I see I think. In either case of the conditional expression either `responseSubscriber` and `pendingResponseSubscriber` are read or `responseSubscriber` is read and `responseSubscriber` is read. I was confused for a bit because I thought only one operand is evaluated. But actually its the case that `responseSubscriber` is read twice when `responseSubscriber == null` evaluates to false.

Good suggestion!